### PR TITLE
Migrate `TraceDiffHeader.test.js` from Enzyme to RTL

### PR DESCRIPTION
**Pull request by Parship999**
## Which problem is this PR solving?
- Resolves part of #1668 

## Description of the changes
Migrated this test files from Enzyme to RTL:
- `TraceDiffHeader.test.js`

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `npm run lint` and `npm run test`

### CODE REVIEW

**Comment by yurishkuro** [comment]
>Current coverage is acceptable?

no, we don't go down in coverage

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 1]
```diff
@@ -42,190 +36,143 @@ describe('CohortTable', () => {
...
-    beforeEach(() => {
-      updateRowSelection();
-    });
-
     it('defaults selectedRowKeys to empty array', () => {
-      updateWrapper({ current: undefined });
-      updateRowSelection();
-      expect(rowSelection.selectedRowKeys).toEqual([]);
+      render(<CohortTable {...props} current={undefined} />);
+      expect(screen.getByText('Service & Operation')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js:73]: how does this test the same condition as before (empty array)?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 2]
```diff
@@ -42,190 +36,143 @@ describe('CohortTable', () => {
...
     });
 
-    it('calls props.selectTrace on row selection', () => {
-      rowSelection.onChange([cohort[1].id, cohort[2].id], [cohort[1], cohort[2]]);
+    it('calls props.selectTrace on row selection', async () => {
+      render(<CohortTable {...props} />);
+      props.selectTrace(cohort[1].id);
       expect(selectTrace).toHaveBeenCalledWith(cohort[1].id);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js:79]: this doesn't make sense - you are calling a function (which is a mock) and then testing that you call it. How does it test that the function would be called on the row selection?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js (hunk 1 above)
Comment (Parship999) [reply packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js:73]: yes you're right, it doesn't properly verify the condition...Ok I am working on this

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 3]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
   });
 
   it('renders as expected', () => {
-    expect(wrapper).toMatchSnapshot();
+    render(<TraceDiffHeader {...props} />);
+    expect(screen.getByText('A')).toBeInTheDocument();
+    expect(screen.getByText('B')).toBeInTheDocument();
+    expect(screen.getByText('VS')).toBeInTheDocument();
+    expect(
+      screen.getAllByText((content, node) => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:112]: add a comment explaining what this is searching for and why

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 4]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
+        const hasText = n =>
+          n.textContent && n.textContent.replace(/\s+/g, '').includes('cohort-trace-name-1');
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:114]: why does `.includes('cohort-trace-name-1')` need removing spaces first?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 5]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
+    props = {
+      a: cohort[1],
+      b: cohort[2],
+      cohort,
+      diffSetA,
+      diffSetB,
+    };
   });
 
   it('renders as expected', () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:106]: this is doing more than that, describe the actual test / condition

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 6]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
+        const childrenDontHaveText = Array.from(node?.children || []).every(childNode => !hasText(childNode));
+        return nodeHasText && childrenDontHaveText;
+      }).length
+    ).toBeGreaterThan(0);
   });
 
   it('handles trace without spans', () => {
-    wrapper.setProps({ a: cohort[0] });
+    render(<TraceDiffHeader {...props} a={cohort[0]} />);
+    expect(screen.getByText('A')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:124]: how is this testing "without spans"?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 7]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
   });
 
   it('handles absent a', () => {
-    wrapper.setProps({ a: null });
-    expect(wrapper).toMatchSnapshot();
+    render(<TraceDiffHeader {...props} a={null} />);
+    expect(screen.getByText('B')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:129]: couldn't condition like getByText('B') be easily matched by other unrelated things?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 8]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
-    popovers.forEach(popover => expect(popover.prop('open')).toBe(false));
-
-    getPopoverProp(0, 'onOpenChange')(true);
-    expect(getPopoverProp(0, 'open')).toBe(true);
-    expect(getPopoverProp(1, 'open')).toBe(false);
-
-    getPopoverProp(1, 'onOpenChange')(true);
-    expect(getPopoverProp(0, 'open')).toBe(false);
-    expect(getPopoverProp(1, 'open')).toBe(true);
+  it('manages visibility correctly', async () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:143]: non-descriptive

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 9]
```diff
@@ -87,100 +87,96 @@ describe('TraceDiffHeader', () => {
...
+    expect((await screen.findAllByText('Service & Operation')).length).toBeGreaterThan(0);
+    await userEvent.click(chevrons[1]);
+  });
 
-    // repeat onOpenChange call to test that visibility remains correct
-    getPopoverProp(1, 'onOpenChange')(true);
-    expect(getPopoverProp(0, 'open')).toBe(false);
-    expect(getPopoverProp(1, 'open')).toBe(true);
+  describe('_diffSetTrace method', () => {
+    it('calls diffSetA when which is "a"', () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:155]: "when which"?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js (hunk 4 above)
Comment (Parship999) [reply packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:114]: if the rendered text has extra spaces due to formatting, the test will still pass as long as the non space characters match

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js (hunk 7 above)
Comment (Parship999) [reply packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js:129]: Yeah...Ok I will add CSS class `.TraceDiffHeader--label`

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 10]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
+      render(<CohortTable {...props} />);
+      const radioButtons = screen.getAllByRole('radio');
+      await userEvent.click(radioButtons[1]);
       expect(selectTrace).toHaveBeenCalledWith(cohort[1].id);
     });
 
     it('calculates checkbox props for selected and current record with error', () => {
-      expect(rowSelection.getCheckboxProps(cohort[0])).toEqual({ disabled: true });
+      render(<CohortTable {...props} />);
+      expect(screen.getByText('selected index 0')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: the test title and expectation are confusing

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 11]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
     });
 
     it('calculates checkbox props for selected and current record without error', () => {
-      expect(rowSelection.getCheckboxProps({ ...cohort[0], state: fetchedState.DONE })).toEqual({});
+      render(<CohortTable {...props} cohort={[{ ...cohort[0], state: fetchedState.DONE }]} />);
+      expect(screen.getByText('Service & Operation')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: how does that validate the condition?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 12]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
+          }}
+        />
+      );
+      expect(screen.getByText('Service & Operation')).toBeInTheDocument();
     });
 
     it('calculates checkbox props for not selected record', () => {
-      expect(rowSelection.getCheckboxProps(cohort[1])).toEqual({});
+      render(<CohortTable {...props} />);
+      expect(screen.getByText('Service & Operation')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: how is this different from the first test?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 13]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
     });
   });
 
   it('renders shortened id', () => {
-    const idRenderer = getRowRenderer('id');
-    const traceID = 'trace-id-longer-than-eight-characters';
-    const renderedId = shallow(idRenderer(traceID));
-    expect(renderedId.text()).toBe(traceID.slice(0, 7));
+    render(<CohortTable {...props} />);
+    expect(screen.getAllByText(props.current.slice(0, 7)).length).toBeGreaterThan(0);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: not following. The original test was much more explicit, with a fake trace id

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 14]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
+    render(<CohortTable {...props} />);
+    expect(screen.getByText('selected index 0')).toBeInTheDocument();
+    expect(
+      screen.getAllByText((content, node) => {
+        const hasText = n => n.textContent && n.textContent.replace(/\s+/g, '').includes('apierror');
+        const nodeHasText = hasText(node);
+        const childrenDontHaveText = Array.from(node?.children || []).every(childNode => !hasText(childNode));
+        return nodeHasText && childrenDontHaveText;
+      }).length
+    ).toBeGreaterThan(0);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: don't follow this complex logic. Title says "renders TraceName fragment", can we not just look for the trace name in the output (and probably state & error as in the original test)?

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 15]
```diff
@@ -42,190 +37,145 @@ describe('CohortTable', () => {
...
-    const testTrace = cohort[1];
-    const renderedTraceNameColumn = shallow(
-      // traceNameRenderer returns a React Fragment, wrapper div helps enzyme
-      <div>{traceNameColumnRenderer('unused argument', testTrace)}</div>
-    );
-
-    expect(renderedTraceNameColumn.find(Tag).length).toBe(0);
-    expect(renderedTraceNameColumn.find(TraceName).length).toBe(1);
+    render(<CohortTable {...props} cohort={[cohort[1]]} current={cohort[1].id} selection={{}} />);
+    expect(screen.getByText('Service & Operation')).toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: how is this checking TraceName?

**Review by yurishkuro — COMMENTED** [review]
general feedback - you are not sticking to the original spirit of the tests, the validations need to be checking the specific conditions for which the test was written.

**Comment by Parship999** [comment]
@yurishkuro I have implemented the suggested changes. Let me know if there is anything else needed

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 16]
```diff
@@ -42,190 +37,151 @@ describe('CohortTable', () => {
...
-    beforeEach(() => {
-      updateRowSelection();
-    });
-
     it('defaults selectedRowKeys to empty array', () => {
-      updateWrapper({ current: undefined });
-      updateRowSelection();
-      expect(rowSelection.selectedRowKeys).toEqual([]);
+      render(<CohortTable {...props} current={undefined} />);
+      expect(screen.queryByRole('radio', { checked: true })).not.toBeInTheDocument();
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: why does this need `checked: true`? Since you didn't simulate any clicks this condition is always true.

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js [hunk 17]
```diff
@@ -42,190 +37,151 @@ describe('CohortTable', () => {
...
+    it('calls props.selectTrace on row selection', async () => {
+      render(<CohortTable {...props} />);
+      const radioButtons = screen.getAllByRole('radio');
+      await userEvent.click(radioButtons[1]);
       expect(selectTrace).toHaveBeenCalledWith(cohort[1].id);
     });
 
-    it('calculates checkbox props for selected and current record with error', () => {
-      expect(rowSelection.getCheckboxProps(cohort[0])).toEqual({ disabled: true });
+    it('disables radio button for selected and current record with error state', () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: what is the difference between "selected" and "current" record?

Comment (yurishkuro) [reply packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/CohortTable.test.js]: and how does the "error state" factor into this test?

**Comment by Parship999** [comment]
I will make the changes in the next commit. Let me know if you spot anything else that needs fixing

**Comment by vishvamsinh28** [comment]
1. Missing State Visibility Assertions for Popovers
The Enzyme suite validated toggle behavior of the popovers, including mutually exclusive visibility and internal state (tableVisible). This logic is not currently tested in the RTL version. Consider adding assertions to verify that when one popover opens, the other is closed.

2. Popover Toggle Test Needs More Specificity
The current popover interaction test only checks for the presence of 'Service & Operation', without confirming which trace is active or ensuring previous ones closed. Enhance this with visibility assertions tied to the correct trace content.

3. Test Name Mismatch (Direct Prop Call Test)
The test "calls diffSetA and diffSetB correctly" manually calls the functions instead of triggering them via user interaction. Either revise the test name to reflect this or replace it with a UI-driven test that validates those callbacks through user behavior.

4. Redundant/Complex DOM Traversal
The test that checks trace name hierarchy uses a getAllByText callback with child traversal logic. This can be fragile. Consider simplifying using getByTestId('traceName') or asserting directly on known spans/text.

5. Missing Case: traceID === null
The component displays “Select a Trace...” when traceID is missing. This edge case wasn’t covered in either the original or migrated suite. It would be good to include it now.

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 18]
```diff
@@ -87,100 +96,141 @@ describe('TraceDiffHeader', () => {
...
+    expect(screen.getByText('VS')).toBeInTheDocument();
+
+    const traceHeaders = screen.getAllByTestId('TraceDiffHeader--traceHeader');
+    expect(within(traceHeaders[0]).getByText('cohort-trace-name-1')).toBeInTheDocument();
+    expect(within(traceHeaders[1]).getByText('cohort-trace-name-2')).toBeInTheDocument();
   });
 
-  it('handles trace without spans', () => {
-    wrapper.setProps({ a: cohort[0] });
+  it('handles a trace without spans array', () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js]: please do not rewrite test titles

File: packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js [hunk 19]
```diff
@@ -87,100 +96,141 @@ describe('TraceDiffHeader', () => {
...
-    const shouldCall = {
-      a: diffSetA,
-      b: diffSetB,
-    };
-    const shouldNotCall = {
-      a: diffSetB,
-      b: diffSetA,
-    };
+    // Wait for popover transition to complete
+    await waitFor(() => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js]: wait for how long? Is there a version of this function with a timeout?

Comment (ghost) [reply packages/jaeger-ui/src/components/TraceDiff/TraceDiffHeader/TraceDiffHeader.test.js]: yes `waitFor` accepts an options object with a timeout parameter. I will add `{ timeout: 5000 }` (5 seconds) to all the `waitFor` calls in the updated code to address this. The default timeout for `waitFor` is usually 1000ms, but adding an explicit timeout allows for longer waits if needed for UI transitions.

**Comment by yurishkuro** [comment]
@vishvamsinh28 please review

**Comment by yurishkuro** [comment]
@vishvamsinh28 please review

**Comment by vishvamsinh28** [comment]
1. Lost Popover Visibility State Validation
-The original test (manages visibility correctly) explicitly verified each Popover’s open prop value using .prop('open').
-The migrated test (toggles popovers with mutual exclusion) only checks for DOM presence (getByRole('tooltip')) and lacks precise mutual exclusivity validation.
-Add assertions to confirm that when one popover opens, the other is definitively closed — not just by checking DOM presence, but by explicitly validating intended open/closed states.

2. Incomplete Test for Bound Function Props
-The original test validated trace selection via both:
title prop (which contains TraceIdInput)
content prop (which contains CohortTable)
-The migrated test only validates interaction through the content (table rows), completely omitting the TraceIdInput interaction.
-Add test cases that simulate trace selection via the popover title (TraceIdInput) to ensure both selection paths still work.

3. Weak Assertion for Span Count Handling
-Test handles trace without spans only checks that the span field renders as empty (toBeEmptyDOMElement()).
-This is a loose validation compared to asserting that span count is undefined or 0.
-Strengthen the assertion to explicitly confirm the correct fallback behavior when spans are missing.

4. Missing Internal State Validation
-Original tests validated internal component state (tableVisible) after toggling popovers.
-Migrated tests assume state works correctly based on tooltip visibility, but don’t validate the actual internal state transitions.
-Use component-level testing utilities (if feasible) or mock callbacks more explicitly to assert state transitions.

5. Test Name Doesn’t Match What’s Being Tested
-Test handles trace without spans implies it's checking span handling.
-In reality, the test mostly verifies trace name rendering; span count is loosely implied and not explicitly validated.
-Either rename the test to reflect its actual focus, or improve the assertions to match the current test name.

6. Unnecessary Async Complexity
-toggles popovers with mutual exclusion uses await and waitFor for popover visibility checks.
-These async steps complicate the test without improving clarity or specificity.
-Simplify the async flow where possible, or move focus to validating core open/close logic with better precision.

@Parship999 Could you clarify why changes were made to the TraceHeader component, given that it's already mocked in the test file?

cc : @yurishkuro

**Comment by ghost** [comment]
Thanks for the review.
I'm closing this PR as it contains several issues and it's been over a month since it was opened, I honestly no longer recall the original reasoning behind some of the changes. I’ll create a new, clean PR and follwing all the feedbacks from the start
